packages/starlight/utils/navigation.ts

```diff
@@ -446,3 +460,38 @@ function stripExtension(path: string) {
 	const periodIndex = path.lastIndexOf('.');
 	return path.slice(0, periodIndex > -1 ? periodIndex : undefined);
 }
+
+/** Get a sidebar badge for a given item. */
+function getSidebarBadge(
+	config: I18nBadgeConfig,
+	locale: string | undefined,
+	itemLabel: string
+): Badge | undefined {
+	if (!config) return;
+	if (typeof config === 'string') {
+		return { variant: 'default', text: config };
+	}
+	return { ...config, text: getSidebarBadgeText(config.text, locale, itemLabel) };
+}
+
+/** Get the badge text for a sidebar item. */
+function getSidebarBadgeText(
+	text: I18nBadge['text'],
+	locale: string | undefined,
+	itemLabel: string
+): string {
+	if (typeof text === 'string') return text;
+	const defaultLang =
+		config.defaultLocale?.lang || config.defaultLocale?.locale || BuiltInDefaultLocale.lang;
+	const defaultText = text[defaultLang];
+
+	if (!defaultText) {
+		throw new AstroError(
+			`The badge text for "${itemLabel}" must have a key for the default language "${defaultLang}".`,
+			'Update the Starlight config to include a badge text for the default language.\n' +
+				'Learn more about sidebar badges internationalization at https://starlight.astro.build/guides/sidebar/#internationalization-with-badges'
+		);
+	}
```

Comment:
Noting that one way to avoid needing this would be to use the same API we use for sidebar items themselves:

```js
{
  label: 'Andromeda',
  translations: { fr: 'Andromède' },
}
```

I don’t have a strong opinion — we use the `string | Record<string, string>` form for the site title, but the `label`/`translations` form here in the sidebar currently. I’m pretty sure I opted for the separate `translations` property when designing the sidebar API to enforce the presence of a default label at the schema level, but having `translations` keys proliferating everywhere also isn’t amazing.

Happy to keep this as is, but wanted to note this while approving!